junit-platform-engine/src/main/java/org/junit/platform/engine/support/store/NamespacedHierarchicalStore.java

```diff
@@ -0,0 +1,358 @@
+	 *
+	 * <p>The {@link CloseAction} will <em>not</em> be called for the removed
+	 * value.
+	 *
+	 * @return previously stored value; may be {@code null}
+	 * @throws NamespacedHierarchicalException if the stored value cannot be cast
+	 *                                         to the required type
+	 */
+	public <T> T remove(N namespace, Object key, Class<T> requiredType) {
+		Object value = remove(namespace, key);
+		return castToRequiredType(key, value, requiredType);
+	}
+
+	private StoredValue storedValue(Supplier<Object> value) {
+		return new StoredValue(insertOrderSequence.getAndIncrement(), value);
+	}
+
+	private StoredValue getStoredValue(CompositeKey<N> compositeKey) {
+		StoredValue storedValue = storedValues.get(compositeKey);
+		if (storedValue != null) {
+			return storedValue;
+		}
+		if (parentStore != null) {
+			return parentStore.getStoredValue(compositeKey);
+		}
+		return null;
+	}
+
+	@SuppressWarnings("unchecked")
+	private <T> T castToRequiredType(Object key, Object value, Class<T> requiredType) {
+		if (value == null) {
+			return null;
+		}
+		if (isAssignableTo(value, requiredType)) {
+			if (requiredType.isPrimitive()) {
+				return (T) getWrapperType(requiredType).cast(value);
+			}
+			return requiredType.cast(value);
+		}
+		// else
+		throw new NamespacedHierarchicalException(
+			String.format("Object stored under key [%s] is not of required type [%s]", key, requiredType.getName()));
+	}
+
+	private static class CompositeKey<N> {
+
+		private final N namespace;
+		private final Object key;
+
+		private CompositeKey(N namespace, Object key) {
+			this.namespace = namespace;
+			this.key = key;
+		}
+
+		@Override
+		public boolean equals(Object o) {
+			if (this == o) {
+				return true;
+			}
+			if (o == null || getClass() != o.getClass()) {
+				return false;
+			}
+			CompositeKey<?> that = (CompositeKey<?>) o;
+			return this.namespace.equals(that.namespace) && this.key.equals(that.key);
+		}
+
+		@Override
+		public int hashCode() {
+			return Objects.hash(namespace, key);
+		}
+
+	}
+
+	private static class StoredValue {
+
+		private final int order;
+		private final Supplier<Object> supplier;
+
+		StoredValue(int order, Supplier<Object> supplier) {
+			this.order = order;
+			this.supplier = supplier;
+		}
+
+		private EvaluatedValue evaluateSafely() {
+			try {
+				return new EvaluatedValue(order, evaluate());
+			}
+			catch (Throwable t) {
+				UnrecoverableExceptions.rethrowIfUnrecoverable(t);
+				return null;
+			}
+		}
+
+		private Object evaluate() {
+			return supplier.get();
+		}
+
+		static Object evaluateIfNotNull(StoredValue value) {
+			return value != null ? value.evaluate() : null;
+		}
+
+	}
+
+	private static class EvaluatedValue {
+
+		private final int order;
+		private final Object value;
+
+		EvaluatedValue(int order, Object value) {
+			this.order = order;
+			this.value = value;
+		}
+	}
+
+	private static class MemoizingSupplier implements Supplier<Object> {
```

Comment:
I think it would be good to add some comments based on [this commit description](https://github.com/junit-team/junit5/commit/9c9085033bf45caf704c7653bfe5d1903f072455) to explain why the complicated `Supplier` construct is used instead of just storing the plain value.

Reply:
Done in 75bb8463c3529f86a8de1d11db2e3fde6b739543.